### PR TITLE
FIX: reject unknown segwit versions as invalid address

### PR DESCRIPTION
**Pull request by Overtorment**
@junderw can you pls take a look

### CODE REVIEW

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/j8x1m4yfezgw4vxu033uvpcn0r?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/9182)

File: class/wallets/legacy-wallet.ts [hunk 1]
```diff
@@ -499,13 +499,22 @@ export class LegacyWallet extends AbstractWallet {
...
    */
   isAddressValid(address: string): boolean {
     try {
-      bitcoin.address.toOutputScript(address);
-      return true;
+      bitcoin.address.toOutputScript(address); // throws, no?
+
+      if (!address.toLowerCase().startsWith('bc1')) return true;
+      const decoded = bitcoin.address.fromBech32(address);
+      return decoded.version <= 1;
```
Comment (junderw) [inline class/wallets/legacy-wallet.ts:516]: Segwit v0 checks are contained within `toOutputScript` so you are fine.

But you should also add some v1 checks.

1. The decoded data is 32 bytes exactly.
2. The 32 bytes is less than the field modulo.
3. The 32 bytes is a valid X point on the curve.

This check can be done easily by prepending 0x02 and using isPoint from some elliptic library.

```js
// example
const ecc = require('tiny-secp256k1'); // I notice you have 1.1.6 in dependencies
const decoded = bitcoin.address.fromBech32(address);
if (decoded.version > 1) return false;
if (decoded.version === 1 && !ecc.isPoint(Buffer.concat([
  Buffer.from([2]),
  decoded.data,
]))) return false;

return true;
```

**Review by junderw — CHANGES_REQUESTED** [review]
Should have more strict checks on v1 segwit.

But tbh, even the current PR is more strict than current, which is better than current.

File: class/wallets/legacy-wallet.ts (hunk 1 above)
Comment (Overtorment) [reply class/wallets/legacy-wallet.ts:516]: i thought lenght checks are already embedded? those asserts in my test case work fine:


```
    assert.ok(!w.isAddressValid('bc1pw5dgrnzv')); // v1, data length != 32
    assert.ok(!w.isAddressValid('bc1p0xlxvlhemja6c4dqv22uapctqupfhlxm9h8z3k2e72q4k9hcz7v8n0nx0muaewav253zgeav')); // v1, data length != 32
```

Comment (junderw) [reply class/wallets/legacy-wallet.ts:516]: Those are length 1 and 41. They are throwing due to the vague segwit rules.

Minimum and maximum segwit length is 2 and 40.

Your wallet will parse a 2 byte data or 40 byte data as a valid v1 address.

Comment (junderw) [reply class/wallets/legacy-wallet.ts:516]:
```
> address.fromBech32('bc1p0xlxvlhemja6c4dqv22uapctqupfhlxm9h8z3k2e72q4k9hcz7v8n0nx0muaewav253zgeav').data.length
41
> address.fromBech32('bc1pw5dgrnzv').data.length
1
```

Comment (junderw) [reply class/wallets/legacy-wallet.ts:516]: https://github.com/bitcoinjs/bitcoinjs-lib/blob/93af5afe67dbc491e36bdfc8d48a00179093f7d1/ts_src/address.ts#L21-L22

Comment (junderw) [reply class/wallets/legacy-wallet.ts:516]: https://github.com/bitcoinjs/bitcoinjs-lib/blob/93af5afe67dbc491e36bdfc8d48a00179093f7d1/ts_src/address.ts#L35-L39

Comment (Overtorment) [reply class/wallets/legacy-wallet.ts]: fixed

**Comment by GladosBlueWallet** [comment]
♫ This was a triumph. I'm making a note here: HUGE SUCCESS ♫

 [android in browser] https://appetize.io/app/59z9y42d3h4xgtz37d6hbmung0?device=pixel4

[download apk](https://lambda-download-android-build.herokuapp.com/download/9184)

**Review by junderw — APPROVED** [review]
LGTM. Here are some extra failures.

```js
'bc1pllllllllllllllllllllllllllllllllllllllllllllallllscqlhrddu' // X is modulo P + 1 (invalid X, but 1 is valid, testing if wrapped modulo (P+1 mod P === 1) will pass)
'bc1pllllllllllllllllllllllllllllllllllllllllllllallllshqcgyklh' // X is modulo P - 1 (invalid X)
'bc1pqtllllllllllllllllllllllllllllllllllllllllllllhlll7zcsqylfl' // data length is 33 (valid point in compressed DER format (33 bytes))
'bc1plllllllllllllllllllllllllllllllllllllllllll0lllu9cegrnmx' // data is length 31 (valid X value with leading 0x00 trimmed)
```

These should all fail the new logic.

**Comment by junderw** [comment]
This is how I generated these cases.
```js
> x.address.toBech32(Buffer.from('fffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc30','hex'), 1, 'bc')
'bc1pllllllllllllllllllllllllllllllllllllllllllllallllscqlhrddu'
> x.address.toBech32(Buffer.from('fffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2e','hex'), 1, 'bc')
'bc1pllllllllllllllllllllllllllllllllllllllllllllallllshqcgyklh'
> x.address.toBech32(Buffer.from('02fffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2c','hex'), 1, 'bc')
'bc1pqtllllllllllllllllllllllllllllllllllllllllllllhlll7zcsqylfl'
> x.address.toBech32(Buffer.from('fffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2e','hex'), 1, 'bc')
'bc1plllllllllllllllllllllllllllllllllllllllllll0lllu9cegrnmx'
```

**Comment by GladosBlueWallet** [comment]
Unbelievable. You, [subject name here], must be the pride of [subject hometown here]!

**Comment by junderw** [comment]
See #4270 for more tests

**Comment by Overtorment** [comment]
thanks! added them in 3195977a0e5176b395f5358b91e8dbfa7327cda1

**Pull request merged**